test/sqllogictest/table_func.slt

```diff
@@ -185,8 +194,18 @@ EXPLAIN PLAN FOR SELECT * FROM x, generate_series(1, a)
 ----
 %0 =
 | Get materialize.public.x (u5)
+| ArrangeBy (#0)
+
+%1 =
+| Get materialize.public.x (u5)
 | FlatMap generate_series(1, #0)
-| | demand = (#0..#2)
+| | demand = (#0, #2)
+
+%2 =
+| Join %0 %1 (= #0 #2)
+| | implementation = Differential %1 %0.(#0)
+| | demand = (#0, #1, #4)
+| Project (#0, #1, #4)
```

Comment:
This is probably a beefy regression of a class we should sort out, somehow. I expect this falls under the general heading of "branch doesn't always need to join the left side again". We should at least decide who's job we want to make doing that, thoug!

Reply:
Same for the subsequent `EXPLAIN` diffs; certainly going to be perf regressions there, pretty substantial to for something like json decoding (`FlatMap` is currently stateless; these joins will break that and have us keep the whole input collection around).